src/wh_server_keystore.c

```diff
@@ -0,0 +1,293 @@
+/* System libraries */
+#include <stdint.h>
+#include <stdlib.h>  /* For NULL */
+#include <string.h>  /* For memset, memcpy */
+
+#include "wolfssl/wolfcrypt/settings.h"
+#include "wolfssl/wolfcrypt/error-crypt.h"
+
+#include "wolfhsm/wh_server.h"
+#include "wolfhsm/wh_server_keystore.h"
+#include "wolfhsm/wh_message.h"
+#include "wolfhsm/wh_packet.h"
+
+int hsmGetUniqueId(whServerContext* server)
+{
+    int i;
+    int ret = 0;
+    uint16_t id;
+    /* try every index, unless both cache and nvm are full one will work */
+    for (id = 1; id < WOLFHSM_NUM_RAMKEYS + 1; id++) {
+        /* check against chache keys */
+        for (i = 0; i < WOLFHSM_NUM_RAMKEYS; i++) {
+            if (id == server->cache[i].meta->id)
+                break;
+        }
+        /* try again if match */
+        if (i < WOLFHSM_NUM_RAMKEYS)
+            continue;
+#if 0
+        /* check against nvm keys */
+        for (i = 0; i < WOLFHSM_KEYSLOT_COUNT && ret == 0; i++) {
+            if (ctx->nvmMetaCache[i].id == id) {
+                break;
+            }
+        }
+        /* try again if match */
+        if (i < WOLFHSM_KEYSLOT_COUNT)
+            continue;
+#endif
+        /* break if our id didn't match either cache or nvm */
+        break;
+    }
+    /* if we've run out of space, return MEMORY_E */
+    if (ret == 0 && id >= WOLFHSM_NUM_RAMKEYS + 1)
+        ret = MEMORY_E;
+    /* ultimately, return found id */
+    if (ret == 0)
+        ret = id;
+    return ret;
+}
+
+int hsmCacheKey(whServerContext* server, whNvmMetadata* meta, uint8_t* in)
+{
+    int i;
+    int foundIndex = -1;
+    /* make sure id is valid */
+    if (meta->id == 0)
+        return BAD_FUNC_ARG;
+    /* make sure the key fits in a slot */
+    if (meta->len > WOLFHSM_NVM_MAX_OBJECT_SIZE)
+        return BUFFER_E;
+    for (i = 0; i < WOLFHSM_NUM_RAMKEYS; i++) {
+        /* check for empty slot or rewrite slot */
+        if (foundIndex == -1 && (server->cache[i].meta->id == 0 ||
```

Comment:
use the macro, not the magic number here
```suggestion
        if (foundIndex == -1 && (server->cache[i].meta->id == WOLFHSM_ID_ERASED ||
```